File: resilient/resilient-annotation-processor/src/main/java/io/koraframework/resilient/annotation/processor/aop/RateLimitKoraAspect.java
package io.koraframework.resilient.annotation.processor.aop;

import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.CodeBlock;
import io.koraframework.annotation.processor.common.MethodUtils;
import io.koraframework.annotation.processor.common.ProcessingErrorException;
import io.koraframework.aop.annotation.processor.KoraAspect;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ExecutableElement;
import java.util.List;
import java.util.Optional;
import java.util.Set;

import static com.palantir.javapoet.CodeBlock.joining;

public class RateLimitKoraAspect implements KoraAspect {

    private static final ClassName ANNOTATION_TYPE = ClassName.get("io.koraframework.resilient.ratelimiter.annotation", "RateLimit");
    private static final ClassName EXCEEDED_EXCEPTION = ClassName.get("io.koraframework.resilient.ratelimiter", "RateLimitExceededException");

    private final ProcessingEnvironment env;

    public RateLimitKoraAspect(ProcessingEnvironment env) {
        this.env = env;
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return Set.of(ANNOTATION_TYPE.canonicalName());
    }

    @Override
    public Set<ClassName> getSupportedAnnotationClassNames() {
        return Set.of(ANNOTATION_TYPE);
    }

    @Override
    public ApplyResult apply(ExecutableElement method, String superCall, AspectContext aspectContext) {
        if (MethodUtils.isPublisher(method) || MethodUtils.isFuture(method)) {
            throw new ProcessingErrorException("Future and Publisher methods are not supported", method);
        }

        final Optional<? extends AnnotationMirror> mirror = method.getAnnotationMirrors().stream()
            .filter(a -> a.getAnnotationType().toString().equals(ANNOTATION_TYPE.canonicalName()))
            .findFirst();

        final String rateLimiterName = mirror.flatMap(a -> a.getElementValues().entrySet().stream()
                .filter(e -> e.getKey().getSimpleName().contentEquals("value"))
                .map(e -> String.valueOf(e.getValue().getValue()))
                .findFirst())
            .orElseThrow();

        var managerType = env.getTypeUtils().getDeclaredType(
            env.getElementUtils().getTypeElement("io.koraframework.resilient.ratelimiter.RateLimiterManager"));
        var fieldManager = aspectContext.fieldFactory().constructorParam(managerType, List.of());
        var rateLimiterType = env.getTypeUtils().getDeclaredType(
            env.getElementUtils().getTypeElement("io.koraframework.resilient.ratelimiter.RateLimiter"));
        var fieldRateLimiter = aspectContext.fieldFactory().constructorInitialized(rateLimiterType,
            CodeBlock.of("$L.get($S)", fieldManager, rateLimiterName));

        return new ApplyResult.MethodBody(buildBodySync(method, superCall, fieldRateLimiter));
    }

    private CodeBlock buildBodySync(ExecutableElement method, String superCall, String rlField) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);
        final CodeBlock methodCall = MethodUtils.isVoid(method)
            ? superMethod
            : CodeBlock.of("var _result = $L", superMethod.toString());

        final CodeBlock returnCall = MethodUtils.isVoid(method)
            ? CodeBlock.of("return")
            : CodeBlock.of("return _result");

        return CodeBlock.builder().add("""
            try {
                $L.acquire();
                $L;
                $L;
            } catch ($T _e) {
                throw _e;
            } catch (Throwable _e) {
                throw _e;
            }
            """, rlField, methodCall.toString(), returnCall.toString(), EXCEEDED_EXCEPTION).build();
    }

    private CodeBlock buildMethodCall(ExecutableElement method, String call) {
        return method.getParameters().stream()
            .map(p -> CodeBlock.of("$L", p))
            .collect(joining(", ", call + "(", ")"));
    }
}

File: resilient/resilient-symbol-processor/src/main/kotlin/io/koraframework/resilient/symbol/processor/aop/RateLimitKoraAspect.kt
package io.koraframework.resilient.symbol.processor.aop

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.MemberName
import com.squareup.kotlinpoet.ksp.toTypeName
import io.koraframework.aop.symbol.processor.KoraAspect
import io.koraframework.ksp.common.AnnotationUtils.findAnnotation
import io.koraframework.ksp.common.AnnotationUtils.findValue
import io.koraframework.ksp.common.CommonClassNames
import io.koraframework.ksp.common.FunctionUtils.isCompletionStage
import io.koraframework.ksp.common.FunctionUtils.isFlow
import io.koraframework.ksp.common.FunctionUtils.isFlux
import io.koraframework.ksp.common.FunctionUtils.isFuture
import io.koraframework.ksp.common.FunctionUtils.isMono
import io.koraframework.ksp.common.FunctionUtils.isVoid
import io.koraframework.ksp.common.exception.ProcessingErrorException
import java.util.concurrent.CompletionStage
import java.util.concurrent.Future

class RateLimitKoraAspect(val resolver: Resolver) : KoraAspect {

    companion object {
        private val ANNOTATION_TYPE = ClassName("io.koraframework.resilient.ratelimiter.annotation", "RateLimited")
        private val RATE_LIMITER = ClassName("io.koraframework.resilient.ratelimiter", "RateLimiter")
        private val EXCEEDED_EXCEPTION = ClassName("io.koraframework.resilient.ratelimiter.exception", "RateLimitExceededException")
    }

    override fun getSupportedAnnotationTypes(): Set<String> {
        return setOf(ANNOTATION_TYPE.canonicalName)
    }

    override fun apply(ksFunction: KSFunctionDeclaration, superCall: String, aspectContext: KoraAspect.AspectContext): KoraAspect.ApplyResult {
        if (ksFunction.isFuture()) {
            throw ProcessingErrorException("@RateLimited can't be applied for types assignable from ${Future::class.java}", ksFunction)
        } else if (ksFunction.isCompletionStage()) {
            throw ProcessingErrorException("@RateLimited can't be applied for types assignable from ${CompletionStage::class.java}", ksFunction)
        } else if (ksFunction.isMono()) {
            throw ProcessingErrorException("@RateLimited can't be applied for types assignable from ${CommonClassNames.mono}", ksFunction)
        } else if (ksFunction.isFlux()) {
            throw ProcessingErrorException("@RateLimited can't be applied for types assignable from ${CommonClassNames.flux}", ksFunction)
        }

        val rateLimiterType = ksFunction.findAnnotation(ANNOTATION_TYPE)!!
            .findValue<KSType>("value")!!
        val baseRateLimiter = resolver.getClassDeclarationByName(RATE_LIMITER.canonicalName)!!.asStarProjectedType()
        if (!baseRateLimiter.isAssignableFrom(rateLimiterType)) {
            throw ProcessingErrorException("@RateLimited value must extend ${RATE_LIMITER.canonicalName}", ksFunction)
        }
        val fieldRateLimiter = aspectContext.fieldFactory.constructorParam(
            rateLimiterType.toTypeName(),
            listOf()
        )

        val body = if (ksFunction.isFlow()) {
            buildBodyFlow(ksFunction, superCall, fieldRateLimiter)
        } else {
            buildBodySync(ksFunction, superCall, fieldRateLimiter)
        }
        return KoraAspect.ApplyResult.MethodBody(body)
    }

    private fun buildBodySync(
        method: KSFunctionDeclaration, superCall: String, fieldRateLimiter: String
    ): CodeBlock {
        val superMethod = buildMethodCall(method, superCall)
        val methodCall = if (method.isVoid()) superMethod else CodeBlock.of("val t = %L", superMethod)
        val returnCall = if (method.isVoid()) CodeBlock.of("") else CodeBlock.of("t")

        return CodeBlock.builder().add(
            """
            return try {
                %L.acquire()
                %L
                %L
            } catch (e: %T) {
                throw e
            } catch (e: Throwable) {
                throw e
            }
            """.trimIndent(),
            fieldRateLimiter, methodCall, returnCall, EXCEEDED_EXCEPTION
        ).build()
    }

    private fun buildBodyFlow(
        method: KSFunctionDeclaration, superCall: String, fieldRateLimiter: String
    ): CodeBlock {
        val flowMember = MemberName("kotlinx.coroutines.flow", "flow")
        val emitMember = MemberName("kotlinx.coroutines.flow", "emitAll")
        val superMethod = buildMethodCall(method, superCall)
        return CodeBlock.builder().add(
            """
            return %M {
                try {
                    %L.acquire()
                    %M(%L)
                } catch (e: %T) {
                    throw e
                } catch (e: Throwable) {
                    throw e
                }
            }
            """.trimIndent(),
            flowMember, fieldRateLimiter, emitMember, superMethod.toString(), EXCEEDED_EXCEPTION
        ).build()
    }

    private fun buildMethodCall(method: KSFunctionDeclaration, call: String): CodeBlock {
        return CodeBlock.of(method.parameters.asSequence().map { p -> CodeBlock.of("%L", p) }.joinToString(", ", "$call(", ")"))
    }
}
